### PR TITLE
Fixed #34980 -- Changed migration operation dependencies to namedtuples.

### PR DESCRIPTION
**Pull request by felixxm**
I've been thinking about changing this for some time, and I'm [not the only one](https://github.com/django/django/pull/16905#discussion_r1306500610), so I gave it a try.

### CODE REVIEW

**Comment by felixxm** [comment]
> Should it have a ticket for posterity's sake?

Sure we can, ticket-34980.

**Comment by carltongibson** [comment]
Great thanks. LGTM but I'm assuming others will want to comment. (So I'll hold off on a ✅ momentarily 😅)

**Comment by felixxm** [comment]
buildbot, test on oracle.

**Review by claudep — APPROVED** [review]
Indeed, looks so much better!

File: django/db/migrations/autodetector.py
Comment (pauloxnet) [inline django/db/migrations/autodetector.py:31]: Could it be possible to use more explicit names without abbreviations?

**Review by pauloxnet — COMMENTED** [review]
A geat improvement! I left a comment

**Comment by nessita** [comment]
I'm very much in favor of this proposal, though, despite these changes being in "internal implementation code", I do wonder if we should keep some compatibility with the former (internal) API, particularly for the `type` field.

Would it be worth overriding `__getitem__` so third party apps that rely on implementation details would not break? Untested:

```python
class OperationDependency(namedtuple(
    "OperationDependency", "app_label model_name field_name type"
)):

    def __getitem__(self, n):
        result = super().__getitem__(n)
        if n == 3:
            if result == DependencyType.CREATE:
                result = True
            elif result == DependencyType.REMOVE:
                result = False
            # etc.
        return result
```

**Comment by felixxm** [comment]
> Would it be worth overriding `__getitem__` so third party apps that rely on implementation details would not break? Untested:

Do you mean a 3rd-party autodetector? I don't think it's worth keeping the previous (quite confusing) map of types (`True`, `False`, `"alter"`, etc.)

File: django/db/migrations/autodetector.py
Comment (pauloxnet) [reply django/db/migrations/autodetector.py:31]: what about `ORDER_WRT_UNSET`? can we use a more explicit name?

Comment (ngnpope) [reply django/db/migrations/autodetector.py:31]: I also think that "foo" isn't all that great here. It was clearly a way of trying to express both "index_together" and "unique_together", but I'm not sure why we didn't just end up with "index_or_unique_together".

I was wondering whether we'd be better with:

```python
    REMOVE_ORDER_WITH_RESPECT_TO = 3
    ALTER_INDEX_OR_UNIQUE_TOGETHER = 4
```

It feels as though these would sync up better with the operation classes: `AlterIndexTogether`, `AlterUniqueTogether`, and `AlterOrderWithRespectTo`. (Also just use "remove" instead of "unset" and "alter" consistently instead of "change".)

File: django/db/migrations/autodetector.py
Comment (ngnpope) [inline django/db/migrations/autodetector.py:36]: I'm not sure we really need to copy this around? It feels like we're trying to fight Python's default everywhere - in the end, this format that _was_ proposed for Python 3.10 and reverted didn't materialise and something slightly different happened in the form of `ReprEnum` in 3.11. There is something to say about the backward compatibility now in `TextChoices`, etc. but I think we shouldn't continue this elsewhere. And I also don't think we actually care about this here as we're only comparing constant values.

```suggestion
```

File: django/db/migrations/autodetector.py
Comment (ngnpope) [inline django/db/migrations/autodetector.py:23]: I think we can simplify things further below by doing something like the following:

```python
class OperationDependency(
    namedtuple("OperationDependency", "app_label model_name field_name type")
):
    @cached_property
    def model_name_lower(self):
        return self.model_name.lower()

    @cached_property
    def field_name_lower(self):
        return self.field_name.lower()
```

This would allow changes such as the following below:

```diff
-                and operation.model_name_lower == dependency.model_name.lower()
-                and operation.name_lower == dependency.field_name.lower()
+                and operation.model_name_lower == dependency.model_name_lower
+                and operation.name_lower == dependency.field_name_lower
```

I'm not sure how often these comparisons occur, but it might save something?

It may also be worth nesting `DependencyType` as it's strictly related to `OperationDependency`:

```python
class OperationDependency(...):
    class Type(Enum):
        ...
    ...
```

And then use `OperationDependency.Type.CREATE`, etc?

File: django/db/migrations/autodetector.py
Comment (ngnpope) [inline django/db/migrations/autodetector.py:708]: This was only formatted like this by black because of the addition.

It now makes sense to flatten it since you're editing the lot:
 ```suggestion
                        *resolve_relation(primary_key_rel, app_label, model_name),
```

**Review by ngnpope — COMMENTED** [review]
This is certainly much nicer to follow 👏🏻

File: django/db/migrations/autodetector.py
Comment (felixxm) [reply django/db/migrations/autodetector.py:31]: > It feels as though these would sync up better with the operation classes: AlterIndexTogether, AlterUniqueTogether, and AlterOrderWithRespectTo. (Also just use "remove" instead of "unset" and "alter" consistently instead of "change".)

But would not sync with autodetector methods and variables.

File: django/db/migrations/autodetector.py
Comment (felixxm) [reply django/db/migrations/autodetector.py:36]: Agreed, removed.

File: django/db/migrations/autodetector.py
Comment (felixxm) [reply django/db/migrations/autodetector.py]: I changed to `REMOVE_` and `ALTER_` prefixes.

Comment (shangxiao) [reply django/db/migrations/autodetector.py]: I guess I don't mind "foo" as we've used it in places like get_foo_display() 🤷‍♂️  It's a little weird but seems somewhat conventional.

File: django/db/migrations/autodetector.py
Comment (felixxm) [reply django/db/migrations/autodetector.py:23]: Updated.

**Comment by felixxm** [comment]
Thanks y'all for reviews :gift: I pushed edits.

File: django/db/migrations/autodetector.py
Comment (ngnpope) [reply django/db/migrations/autodetector.py]: I'm not going to lose sleep over it 😄 With the change to the suggested prefixes the "foo" ain't so bad.

**Review by ngnpope — APPROVED** [review]
Fab! Nothing more from me.

_(Haha. Just a few seconds too late...)_

**Pull request merged**
